src/vmm/src/vstate/vcpu/x86_64.rs

```diff
@@ -291,9 +307,55 @@ impl KvmVcpu {
             vcpu_events,
             xcrs,
             xsave,
+            tsc_khz,
         })
     }
 
+    // Scale the TSC frequency of this vCPU to the one provided as a parameter.
+    pub fn set_tsc_khz(&self, state_tsc: Option<u32>, cpuid: &CpuId) -> Result<()> {
+        // Check if CPU models are the same.
+        if let Ok(curr_cpuid) = cpuid::common::get_cpuid(0x1, 0) {
+            if is_same_model(cpuid, curr_cpuid.eax) {
+                // If they are the same, we don't need to do anything else.
+                return Ok(());
+            }
+        }
+
+        // Try to get a reading of the current TSC freq.
+        let curr_kvm_tsc = self.fd.get_tsc_khz();
+
+        // Extract the TSC freq from the state.
+        let snapshot_tsc_freq = if state_tsc.is_none() {
+            // No TSC freq in snapshot means we have to fail-fast.
+            return Err(Error::TSCConfiguration);
+        } else {
+            state_tsc.unwrap()
+        };
+
+        match curr_kvm_tsc {
+            Ok(curr_tsc_freq) => {
+                // Compare the current TSC freq to the one found
+                // in the state. If they are different, we need to
+                // scale the TSC to the freq found in the state.
+                if curr_tsc_freq != snapshot_tsc_freq {
```

Comment:
I think we should allow for some margin of error here, since even on the same CPU you might get different values during calibration.

Reply:
Any margin of error here can lead to a great divergence in TSC over a period of time.

Reply:
Some differences can be corrected by the guest NTP, more info in QEMU approach: https://patchew.org/QEMU/20200616165805.GA324612@fuller.cnet/

Reply:
Updated it/